File: 백준/1000/1012/1012.js
// https://www.acmicpc.net/problem/1012
const fs = require("fs");
const stdin = (
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString()
    : `1
    5 3 6
    0 2
    1 2
    2 2
    3 2
    4 2
    4 0
`
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const C = +input();

for (let i = 0; i < C; i++) {
  const [M, N, K] = input()
    .trim()
    .split(" ")
    .map((v) => +v);

  const arr = new Array(N).fill(null).map((_) => new Array(M).fill(0));
  for (let j = 0; j < K; j++) {
    const [Y, X] = input()
      .trim()
      .split(" ")
      .map((v) => +v);
    arr[X][Y] = 1;
  }
  let count = 0;
  const bfs = (A, B) => {
    const isValid = (x, y) => 0 <= x && x < N && 0 <= y && y < M;
    const arrows = [
      [0, 1],
      [0, -1],
      [1, 0],
      [-1, 0],
    ];
    // console.log(A, B);
    arr[A][B] = 0;
    const q = [[A, B]];
    while (q.length > 0) {
      const [x, y] = q.shift();
      for (let [xx, yy] of arrows) {
        const [xxx, yyy] = [x + xx, y + yy];
        if (isValid(xxx, yyy) && arr[xxx][yyy] === 1) {
          arr[xxx][yyy] = 0;
          q.push([xxx, yyy]);
        }
      }
    }
  };
  for (let k = 0; k < N; k++) {
    for (let l = 0; l < M; l++) {
      if (arr[k][l] === 1) {
        bfs(k, l);
        count++;
      }
    }
  }
  console.log(count);
}

File: 백준/1012.py
# https://www.acmicpc.net/problem/1012


import sys
from collections import deque


def bfs(k, j):
    q = deque([[k, j]])
    while q:
        x, y = q.popleft()
        arr[x][y] = 0
        for xx, yy in [[1, 0], [0, 1], [-1, 0], [0, -1]]:
            xxx, yyy = xx+x, yy+y
            if 0 <= xxx < N and 0 <= yyy < M and arr[xxx][yyy] == 1:
                arr[xxx][yyy] = 0
                q.append([xxx, yyy])
    return 1


C = int(sys.stdin.readline().strip())
q = deque([])
for i in range(C):
    M, N, K = map(int, sys.stdin.readline().strip().split())
    # print(M, N, K)
    arr = [[0] * M for _ in range(N)]
    for k in range(K):
        m, n = map(int, sys.stdin.readline().strip().split())
        arr[n][m] = 1
    # print(arr)
    result = 0
    for k in range(N):
        for j in range(M):
            if arr[k][j] == 1:
                result += bfs(k, j)
    print(result)
